Error when trying to implement turtlesim node

On ubuntu 20.4 with ROS noetic install.

Tried making the following changes to basic_publisher.rs example
```rs
+ use geometry_msgs::{Twist, Vector3};
- roslibrust_codegen_macro::find_and_generate_ros_messages!("roslibrust/local_msgs");
+ roslibrust_codegen_macro::find_and_generate_ros_messages!();

-     let publisher = client.advertise::<std_msgs::Header>("talker").await?;
+    let publisher = client.advertise::<Twist>("/turtle1/cmd_vel").await?;
    loop {
-         let msg = std_msgs::Header::default();
+        let msg = Twist {
+            linear: Vector3 {
+                x: 0.0,
+                y: 0.0,
+                z: 0.0,
+            },
+            angular: Vector3 {
+               x: 1.0,
+                y: 0.0,
+                z: 0.0,
+            },
+        };
```
This was the error output from the console after running the following
```
cargo run --example basic_publisher
```

```

error[E0412]: cannot find type `DurationI` in module `std_msgs`
 --> roslibrust/examples/basic_publisher.rs:4:1
  |
4 | roslibrust_codegen_macro::find_and_generate_ros_messages!();
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `std_msgs`
  |
  = note: this error originates in the macro `roslibrust_codegen_macro::find_and_generate_ros_messages` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `TimeI` in module `std_msgs`
 --> roslibrust/examples/basic_publisher.rs:4:1
  |
4 | roslibrust_codegen_macro::find_and_generate_ros_messages!();
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `std_msgs`
  |
  = note: this error originates in the macro `roslibrust_codegen_macro::find_and_generate_ros_messages` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `DurationI` in module `self`
 --> roslibrust/examples/basic_publisher.rs:4:1
  |
4 | roslibrust_codegen_macro::find_and_generate_ros_messages!();
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `self`
  |
  = note: this error originates in the macro `roslibrust_codegen_macro::find_and_generate_ros_messages` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `TimeI` in module `self`
 --> roslibrust/examples/basic_publisher.rs:4:1
  |
4 | roslibrust_codegen_macro::find_and_generate_ros_messages!();
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `self`
  |
  = note: this error originates in the macro `roslibrust_codegen_macro::find_and_generate_ros_messages` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0283]: type annotations needed
 --> roslibrust/examples/basic_publisher.rs:4:1
  |
4 | roslibrust_codegen_macro::find_and_generate_ros_messages!();
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
  |
  = note: cannot satisfy `_: Default`
  = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0283, E0412.
For more information about an error, try `rustc --explain E0283`.
```
The text file would have been a .rs if let me paste that in here, but output.txt is the printed output of the variable "source" from roslibrust_test/main.rs

[output.txt](https://github.com/Carter12s/roslibrust/files/9464700/output.txt)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
  - Failure in message generation caused by files without extensions
  - Failure in message generation caused by failing to quote around string constants
+ - Failure in message generation caused by bad desing of integral types TimeI and DurationI
 
 ## 0.1.0 - August 28, 2022
 Initial public release

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,11 +18,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
  - Support for service servers
+ - Into<> helpers for Time and Duration for tokio and std
 
 ### Fixed
  - Failure in message generation caused by files without extensions
  - Failure in message generation caused by failing to quote around string constants
- - Failure in message generation caused by bad desing of integral types TimeI and DurationI
+ - Failure in message generation caused by bad design of integral types TimeI and DurationI
 
 ## 0.1.0 - August 28, 2022
 Initial public release

--- a/roslibrust/local_msgs/std_msgs/msg/DurationI.msg
+++ b/roslibrust/local_msgs/std_msgs/msg/DurationI.msg
@@ -1,2 +1,0 @@
-int32 sec
-int32 nsec

--- a/roslibrust/local_msgs/std_msgs/msg/TimeI.msg
+++ b/roslibrust/local_msgs/std_msgs/msg/TimeI.msg
@@ -1,2 +1,0 @@
-uint32 secs
-uint32 nsecs

--- a/roslibrust/src/integral_types.rs
+++ b/roslibrust/src/integral_types.rs
@@ -6,10 +6,41 @@ pub struct Time {
     pub nsecs: u32,
 }
 
+impl Into<Time> for std::time::SystemTime {
+    fn into(self) -> Time {
+        let delta = self
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("Failed to convert system time into unix epoch");
+        let downcast_secs = u32::try_from(delta.as_secs()).expect("Failed to convert system time to ROS representation, seconds term overflows u32 likely");
+        Time {
+            secs: downcast_secs,
+            nsecs: delta.subsec_nanos(),
+        }
+    }
+}
+
+// TODO provide chrono conversions here behind a cfg flag
+
 /// Matches the integral ros1 duration type, with extensions for ease of use
-// NOTE: Is not a message in and of itself use std_msgs/Duration for that
+/// NOTE: Is not a message in and of itself use std_msgs/Duration for that
 #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
 pub struct Duration {
     pub sec: i32,
     pub nsec: i32,
 }
+
+/// Note this provides both tokio::time::Duration and std::time::Duration
+impl Into<Duration> for tokio::time::Duration {
+    fn into(self) -> Duration {
+        let downcast_sec = i32::try_from(self.as_secs())
+            .expect("Failed to cast tokio duration to ROS duration, secs could not fit in i32");
+        let downcast_nsec = i32::try_from(self.subsec_nanos())
+            .expect("Failed to cast tokio duration ROS duration, nsecs could not fit in i32");
+        Duration {
+            sec: downcast_sec,
+            nsec: downcast_nsec,
+        }
+    }
+}
+
+// TODO: provide chrono conversions here behind a cfg flag

--- a/roslibrust/src/integral_types.rs
+++ b/roslibrust/src/integral_types.rs
@@ -1,0 +1,15 @@
+/// Matches the integral ros1 type time, with extensions for ease of use
+/// NOTE: Is not a message in and of itself use std_msgs/Time for that
+#[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
+pub struct Time {
+    pub secs: u32,
+    pub nsecs: u32,
+}
+
+/// Matches the integral ros1 duration type, with extensions for ease of use
+// NOTE: Is not a message in and of itself use std_msgs/Duration for that
+#[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
+pub struct Duration {
+    pub sec: i32,
+    pub nsec: i32,
+}

--- a/roslibrust/src/lib.rs
+++ b/roslibrust/src/lib.rs
@@ -16,6 +16,9 @@ pub use subscriber::*;
 mod publisher;
 pub use publisher::*;
 
+// Contains definition of custom ROS types that codegen emits
+pub mod integral_types;
+
 // Tests are fully private module
 mod integration_tests;
 

--- a/roslibrust_codegen/src/parse.rs
+++ b/roslibrust_codegen/src/parse.rs
@@ -16,8 +16,8 @@ lazy_static::lazy_static! {
         ("float32", "f32"),
         ("float64", "f64"),
         ("string", "std::string::String"),
-        ("time", "TimeI"),
-        ("duration", "DurationI"),
+        ("time", "::roslibrust::integral_types::Time"),
+        ("duration", "::roslibrust::integral_types::Duration"),
         ("Header", "Header"),
     ].into_iter().collect();
 }
@@ -151,7 +151,7 @@ pub fn parse_ros_service_file(data: String, name: String, package: &String) -> S
 }
 
 pub fn replace_ros_types_with_rust_types(mut msg: MessageFile) -> MessageFile {
-    const INTERNAL_STD_MSGS: [&str; 3] = ["Header", "TimeI", "DurationI"];
+    const INTERNAL_STD_MSGS: [&str; 1] = ["Header"];
     msg.constants = msg
         .constants
         .into_iter()

--- a/roslibrust_test/src/lib.rs
+++ b/roslibrust_test/src/lib.rs
@@ -14,7 +14,7 @@ pub mod actionlib_msgs {
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct GoalID {
-        pub stamp: std_msgs::TimeI,
+        pub stamp: ::roslibrust::integral_types::Time,
         pub id: std::string::String,
     }
     impl ::roslibrust::RosMessageType for GoalID {
@@ -560,7 +560,7 @@ pub mod nav_msgs {
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct MapMetaData {
-        pub map_load_time: std_msgs::TimeI,
+        pub map_load_time: ::roslibrust::integral_types::Time,
         pub resolution: f32,
         pub width: u32,
         pub height: u32,
@@ -670,7 +670,7 @@ pub mod rosapi {
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct GetTimeResponse {
-        pub data: std_msgs::TimeI,
+        pub data: ::roslibrust::integral_types::Time,
     }
     impl ::roslibrust::RosMessageType for GetTimeResponse {
         const ROS_TYPE_NAME: &'static str = "rosapi/GetTimeResponse";
@@ -1114,7 +1114,7 @@ pub mod sensor_msgs {
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct TimeReference {
         pub header: std_msgs::Header,
-        pub time_ref: std_msgs::TimeI,
+        pub time_ref: ::roslibrust::integral_types::Time,
         pub source: std::string::String,
     }
     impl ::roslibrust::RosMessageType for TimeReference {
@@ -1258,22 +1258,12 @@ pub mod std_msgs {
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct Duration {
-        pub data: self::DurationI,
+        pub data: ::roslibrust::integral_types::Duration,
     }
     impl ::roslibrust::RosMessageType for Duration {
         const ROS_TYPE_NAME: &'static str = "std_msgs/Duration";
     }
     impl Duration {}
-    #[allow(non_snake_case)]
-    #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
-    pub struct DurationI {
-        pub sec: i32,
-        pub nsec: i32,
-    }
-    impl ::roslibrust::RosMessageType for DurationI {
-        const ROS_TYPE_NAME: &'static str = "std_msgs/DurationI";
-    }
-    impl DurationI {}
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct Empty {}
@@ -1323,7 +1313,7 @@ pub mod std_msgs {
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct Header {
         pub seq: u32,
-        pub stamp: self::TimeI,
+        pub stamp: ::roslibrust::integral_types::Time,
         pub frame_id: std::string::String,
     }
     impl ::roslibrust::RosMessageType for Header {
@@ -1439,22 +1429,12 @@ pub mod std_msgs {
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct Time {
-        pub data: self::TimeI,
+        pub data: ::roslibrust::integral_types::Time,
     }
     impl ::roslibrust::RosMessageType for Time {
         const ROS_TYPE_NAME: &'static str = "std_msgs/Time";
     }
     impl Time {}
-    #[allow(non_snake_case)]
-    #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
-    pub struct TimeI {
-        pub secs: u32,
-        pub nsecs: u32,
-    }
-    impl ::roslibrust::RosMessageType for TimeI {
-        const ROS_TYPE_NAME: &'static str = "std_msgs/TimeI";
-    }
-    impl TimeI {}
     #[allow(non_snake_case)]
     #[derive(:: serde :: Deserialize, :: serde :: Serialize, Debug, Default, Clone, PartialEq)]
     pub struct UInt16 {
@@ -1699,7 +1679,7 @@ pub mod trajectory_msgs {
         pub velocities: std::vec::Vec<f64>,
         pub accelerations: std::vec::Vec<f64>,
         pub effort: std::vec::Vec<f64>,
-        pub time_from_start: std_msgs::DurationI,
+        pub time_from_start: ::roslibrust::integral_types::Duration,
     }
     impl ::roslibrust::RosMessageType for JointTrajectoryPoint {
         const ROS_TYPE_NAME: &'static str = "trajectory_msgs/JointTrajectoryPoint";
@@ -1722,7 +1702,7 @@ pub mod trajectory_msgs {
         pub transforms: std::vec::Vec<geometry_msgs::Transform>,
         pub velocities: std::vec::Vec<geometry_msgs::Twist>,
         pub accelerations: std::vec::Vec<geometry_msgs::Twist>,
-        pub time_from_start: std_msgs::DurationI,
+        pub time_from_start: ::roslibrust::integral_types::Duration,
     }
     impl ::roslibrust::RosMessageType for MultiDOFJointTrajectoryPoint {
         const ROS_TYPE_NAME: &'static str = "trajectory_msgs/MultiDOFJointTrajectoryPoint";
@@ -1755,7 +1735,7 @@ pub mod visualization_msgs {
         pub outline_color: std_msgs::ColorRGBA,
         pub filled: u8,
         pub fill_color: std_msgs::ColorRGBA,
-        pub lifetime: std_msgs::DurationI,
+        pub lifetime: ::roslibrust::integral_types::Duration,
         pub points: std::vec::Vec<geometry_msgs::Point>,
         pub outline_colors: std::vec::Vec<std_msgs::ColorRGBA>,
     }
@@ -1890,7 +1870,7 @@ pub mod visualization_msgs {
         pub pose: geometry_msgs::Pose,
         pub scale: geometry_msgs::Vector3,
         pub color: std_msgs::ColorRGBA,
-        pub lifetime: std_msgs::DurationI,
+        pub lifetime: ::roslibrust::integral_types::Duration,
         pub frame_locked: bool,
         pub points: std::vec::Vec<geometry_msgs::Point>,
         pub colors: std::vec::Vec<std_msgs::ColorRGBA>,

